compress-mpq.c

```diff
@@ -557,97 +552,77 @@ void InitCache() {
-    uint32_t path_len = (uint32_t)strlen(sanitized_path);
-    WriteLE32(file, path_len);
-
-    // Write the sanitized file path followed by a newline
-    fwrite(sanitized_path, sizeof(char), path_len, file);
-    fputc('\n', file);
-
-    // Write the content followed by the out buffer
-    fwrite(content, sizeof(unsigned char), insize, file);
-    fwrite(out, sizeof(unsigned char), outsize, file);
-
-    // Close the file
+    int success = 0;
+    if(fwrite(out, sizeof(unsigned char), outsize, file) == outsize) {
+        success = 1;
+    }
     fclose(file);
+
+    if(success == 0) {
+        remove(cache_path);
+        perror("Failed to update cache file");
+    }    
 }
 
-int ReadCache(char *path, size_t insize, unsigned char *content, unsigned char *out, size_t *outsize, uint32_t *flags){
-    // Sanitize the input path by replacing slashes with hyphens
-    char sanitized_path[1024];
-    SanitizePath(path, sanitized_path);
+int ReadCache(char *path, const size_t insize, const unsigned char *content, unsigned char *out, size_t *outsize, uint32_t *flags){
+    char cache_path[1024];
 
-    // Construct the full file path: "{CWD}/cache/{sanitized_path}"
-    char full_path[1024];
-    snprintf(full_path, sizeof(full_path), "./cache/%s", sanitized_path);
+    {
+      char in_archive_path64[971];
+      char content_hash64[45];
+      char content_hash[32];
 
-    // Check if the file exists
-    struct stat st;
-    if (stat(full_path, &st) != 0) {
-        return 0;  // File does not exist
+      // Construct the full file path: "{CWD}/cache/{content_hash64}-{in_archive_path64}"
+      lonesha256(content_hash, content, insize);
+      Base64URLEncode(content_hash64, content_hash, sizeof(content_hash));
+      Base64URLEncode(in_archive_path64, path, strlen(path));
+      snprintf(cache_path, sizeof(cache_path), "./cache/%s-%s", content_hash64, in_archive_path64);
     }
 
-    // Open the file for reading
-    FILE *file = fopen(full_path, "rb");
+    // For reading and writing, only if it exists.
+    FILE *file = fopen(cache_path, "rb+");
     if (!file) {
-        perror("Failed to open cache file");
-        return 0;  // If we cannot open the file, treat it as not existing
+        return 0;
     }
 
-    // Read insize from the file and compare with input
-    uint32_t cached_insize = ReadLE32(file);
-    if (cached_insize != (uint32_t)insize) {
-        fclose(file);
-        return 0;  // insize does not match
-    }
+    fseek(file, 0, SEEK_END);
+    *outsize = ftell(file);
+    fseek(file, 0, SEEK_SET);
 
-    // Read outsize and flags from the file and output them
-    *outsize = ReadLE32(file);
-    *flags = ReadLE32(file);
-
-    // Read the length of the cached file path
-    uint32_t path_len = ReadLE32(file);
+    fread(out, sizeof(unsigned char), *outsize, file);
+    fclose(file);
 
-    // Skip the stored file path and newline
-    fseek(file, path_len + 1, SEEK_CUR);  // Skip file path and newline
+    // TODO:
```

Comment:
I don't consider this a TODO 😛

Reply:
Well. Rather than a risk of collision, a likelier threat is unexpected OS shutdown or power loss, which would corrupt the cache.

Previous design avoided that threat by storing the expected data size inside the cache file, but if we only store the allegedly compressed data, then the risk is on the table.